Clean export tempfiles on failure

esoergel opened the pull request:
Best reviewed :tropical_fish: :heavy_multiplication_x: :tropical_fish: 
@mkangia
@emord @snopoke potentially interested parties

esoergel committed:
Make context manager to clean tempfiles

This is intended to establish a pattern for accessing tempfiles, where there's a
with statement for file reading (like normal), and a with statement for the
existence of the file.

tempfile.TemporaryFile does automatically clean itself up on close, but then if
you're planning to both read and write to the file, you need to `seek` around.
This wrapper lets us open for reads and open for writes like we normally do,
but still have an automatic cleanup step.

One potential issue with this is that you can apparently remove a file without
first closing it.  That could cause weird bugs if something went wrong.  This
util could control file access as well, but that might be overly defensive.  I'm
sticking this approach for now to keep things simple and preserve the standard
python API as much as possible.

esoergel committed:
Convert get_export_docs to use TransientTempfile

esoergel committed:
Don't instantiate ExportFile just for .path

esoergel committed:
Remove unused parameter

esoergel committed:
Extract repetitive test code

esoergel committed:
Wrap get_export_file for tests

To make it easier to change and to remove some boilerplate

esoergel committed:
Manage temp-file at the highest level so it's cleaned up properly

millerdev commented on corehq/util/files.py:
```diff
@@ -40,3 +42,25 @@ def safe_filename_header(filename, extension=None):
+    """
+
+    def __enter__(self):
+        _, self.path = tempfile.mkstemp()
```
I think you want to call `.close()` on thing you're throwing away here (`_`).

esoergel replied:
Oh yikes, you're right.  We do this all over the place.  Updated.

esoergel commented on corehq/apps/export/tests/test_export_form_subcases.py:
```diff
@@ -15,7 +15,7 @@
 from corehq.apps.export.dbaccessors import delete_all_export_data_schemas
 from corehq.apps.export.models import FormExportDataSchema, FormExportInstance
-from corehq.apps.export.export import get_export_file
+from corehq.apps.export.tests.util import get_export_json
```
This commit is best viewed with whitespace ignored:
https://github.com/dimagi/commcare-hq/pull/20659/commits/6d55dfd8576d5f7161534ac4c5b1bc86a40af282?w=1

esoergel commented on corehq/apps/export/export.py:
```diff
@@ -48,19 +47,18 @@ def __enter__(self):
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.file.close()
-        os.remove(self.path)
```
This is the old way these were cleaned up - via `with` statement used when the generated file was read.  Removing this because `TransientTempfile` handles the cleanup now.

emord reviewed (commented):
How hard would it be to change the code to use `tempfile.TemporaryFile` ? From what I can tell the difference is that your helper provides a path and the official temporary file provides a file like object.

(also nit that `TransientTempfile` feels a little redundant)

standard lib: https://docs.python.org/2/library/tempfile.html#tempfile.TemporaryFile

esoergel commented:
> How hard would it be to change the code to use tempfile.TemporaryFile ? From what I can tell the difference is that your helper provides a path and the official temporary file provides a file like object.

I commented about that a bit in my commit message: https://github.com/dimagi/commcare-hq/pull/20659/commits/f34f48e98bd9c054865b920153a3ef2638526444
`NamedTemporaryFile` provides a path and otherwise has the same API as `TemporaryFile`.  The problem is both of them automatically delete the file when it's closed (although you can make `NamedTemporaryFile` persist if you want).  To use one of those, the code would probably use a `with` statement at the same places as done here, and pass around a file object rather than a path.  The file object would be open for reading and writing.  When the writes are finished, it'd need to `seek` back to the start of the file before the reads begin.

In short, that would probably work too, I just thought this presented a nicer interface, since we can read and write in isolation and close the file in between, which felt more in line with how we typically access files.

> also nit that `TransientTempfile` feels a little redundant

Do you mean in name?  I agree, and am open to suggestions!

emord commented:
>  commented about that a bit in my commit message: f34f48e
ack. I think I read that and then immediately forgot about it. That makes sense.

Maybe `ReadableTemporaryFile`? Not the greatest, but it hints at the benefit of using this class over the standard library.

esoergel committed:
Close the unused open file handle

I didn't see anything in tempfile that looked like we could skip opening it in
the first place

millerdev commented on corehq/util/files.py:
```diff
@@ -59,7 +59,8 @@ class TransientTempfile(object):
     def __enter__(self):
-        _, self.path = tempfile.mkstemp()
+        fd, self.path = tempfile.mkstemp()
+        os.close(fd)
```
There's a smell here because this is closing, and then inside the context manager immediately re-opening the same file. From a resource management perspective this might be more efficient:

```py
@attr.s
class TransientTempfile(object):
    """
    Manage a temporary file that can be opened and closed as needed, but will
    be automatically cleaned up on exit.

        with TransientTempfile() as fileobj, path:
            # path exists throughout this block
            with fileobj:
                fileobj.write("Adding stuff to the file")
            with open(path) as f:
                do_stuff(f.read())
        # path no longer exists
    """

    mode = attr.ib(default="wt")
    encoding = attr.ib(default="utf-8")

    def __enter__(self):
        fd, self.path = tempfile.mkstemp()
        self.encoding = None if "b" in self.mode else self.encoding
        fileobj = io.open(fd, self.mode, encoding=self.encoding)
        return fileobj, self.path

    def __exit__(self, exc_type, exc_val, exc_tb):
        os.remove(self.path)
```

Maybe this makes the implementation too complex? Your call if you want to use it or not.

esoergel replied:
I thought about making this util manage opening and closing the tempfile as well, but I like the idea of keeping it responsible solely for allocating the tempfile and not making a new api for file access.  I think I'll keep it as-is for now.

millerdev commented on corehq/apps/export/export.py:
```diff
@@ -336,19 +309,19 @@ def get_export_file(export_instances, filters, progress_tracker=None):
-                yield doc
-        os.remove(temp_path)
+        with TransientTempfile() as temp_path:
+            with open(temp_path, 'w') as f:
```
I think this is using the default `open` when it should be using `io.open` and passing in `encoding="utf-8"`. Same below on line 321.

esoergel replied:
updated

esoergel replied:
Just reverted that change because it caused build failures.  I'm just gonna leave these as they were beforehand and deal with making this code python 3 compatible in another PR.

millerdev commented on corehq/apps/export/multiprocess.py:
```diff
@@ -185,12 +185,15 @@ def _doc_iter():
-    writer = get_export_writer(export_instances, allow_pagination=False)
+    # Multiprocess exports sometimes intentionally keep the tempfile,
+    # so TransientTempfile isn't appropriate here
+    _, temp_path = tempfile.mkstemp()
```
should close file descriptor (`_`)

esoergel replied:
updated

millerdev reviewed (commented):
Didn't do a complete review, but made a few comments related to opening files with text encoding. Not sure if some of this will work on Python 3 without being more explicit about encodings.

esoergel committed:
Use io.open in exports/export.py

esoergel committed:
Don't keep around open file descriptor

mkangia commented on corehq/apps/export/export.py:
```diff
@@ -336,19 +310,19 @@ def get_export_file(export_instances, filters, progress_tracker=None):
+        with TransientTempfile() as temp_path:
+            with open(temp_path, 'w', encoding='utf-8') as f:
+                for doc_id in scroll_result:
+                    f.write(doc_id + '\n')
```
would it be better to write at once?
`f.write('\n'.join(doc_id for doc_id in scroll_result))`

This would probably depend on the number of ids here. if its just a few or too large this might be not-much-difference/too-much-memory but does avoid that many `write` method calls.

esoergel replied:
It's potentially a very big list, and I think that approach would require pulling it all into memory and committing in one go.  Perhaps a chunked approach would be better?  I'll call that out of scope for this PR, since I wasn't intending to change what happens with the tempfiles, just that they are guaranteed to get cleaned up.

mkangia commented on corehq/apps/export/tasks.py:
```diff
@@ -24,27 +24,29 @@
+                mimetype=file_format.mimetype,
+                content_disposition=safe_filename_header(filename, file_format.extension),
+                download_id=download_id,
+            )
```
Can use `ExposeBlobDownload` here to do some of this? Probably out of scope of this PR so feel free to skip it if needed.

mkangia reviewed (approved):
Couple of minor things.

esoergel committed:
Revert "Use io.open in exports/export.py"

This reverts commit c2bd01a46c0af1c6858f2cc0842fb05ec3d4bce9.
